File: cmd/dedup_bench_pass2.go
// file: cmd/dedup_bench_pass2.go
// version: 1.0.0
// guid: 2b3c4d5e-6f7a-8901-bcde-222222222222

//go:build bench

package cmd

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/jdfalk/audiobook-organizer/internal/ai"
	"github.com/jdfalk/audiobook-organizer/internal/config"
	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
	"github.com/spf13/cobra"
)

var (
	pass2ResultsPath string
	pass2GroupsPath  string
	pass2Model       string
	pass2Threshold   string
)

var dedupBenchPass2Cmd = &cobra.Command{
	Use:   "pass2",
	Short: "Second-pass enrichment for uncertain suggestions",
	Long: `Takes medium/low confidence results from a first-pass groups run,
enriches them with book title data from the server, and resubmits
for a more informed decision via batch API.`,
	RunE: runDedupBenchPass2,
}

func init() {
	dedupBenchPass2Cmd.Flags().StringVar(&pass2ResultsPath, "results", "", "Path to first-pass batch_output.jsonl (required)")
	dedupBenchPass2Cmd.Flags().StringVar(&pass2GroupsPath, "groups", "", "Path to groups.json from first pass (required)")
	dedupBenchPass2Cmd.Flags().StringVar(&pass2Model, "model", "gpt-5.1", "Model for second pass")
	dedupBenchPass2Cmd.Flags().StringVar(&pass2Threshold, "threshold", "medium", "Include suggestions at this confidence and below (medium or low)")
	_ = dedupBenchPass2Cmd.MarkFlagRequired("results")
	_ = dedupBenchPass2Cmd.MarkFlagRequired("groups")
}

const pass2SystemPrompt = `You are an expert audiobook metadata reviewer performing a SECOND-PASS verification.

You previously reviewed author groups and flagged some as uncertain. Now you are given:
1. The original group of author name variants
2. Your previous suggestion and reasoning
3. NEW DATA: Book titles for each author variant, so you can determine if they are the same person

Use the book titles to make a more informed decision:
- If both variants have books in the same series or genre, they are likely the same person → merge
- If they have completely unrelated books, they are likely different people → split
- Narrators tend to have books across many unrelated genres/authors
- Publishers have many unrelated books

INITIALS FORMATTING: Always use spaces after periods in initials: "C. B. Lee" not "C.B. Lee".

Return ONLY valid JSON: {"suggestions": [{"group_index": N, "action": "merge|split|rename|skip|alias|reclassify", "canonical_name": "Correct Name", "reason": "brief explanation using book title evidence", "confidence": "high|medium|low", "roles": {"author": {"name": "Name", "variants": ["V1"], "reason": "why"}, "narrator": {"name": "Name", "ids": [], "reason": "why"}, "publisher": {"name": "Name", "ids": [], "reason": "why"}}}]}`

func runDedupBenchPass2(cmd *cobra.Command, args []string) error {
	apiKey := os.Getenv("OPENAI_API_KEY")
	if apiKey == "" {
		config.InitConfig()
		apiKey = config.AppConfig.OpenAIAPIKey
	}
	if apiKey == "" {
		return fmt.Errorf("OPENAI_API_KEY env var required")
	}
	if benchServerURL == "" {
		return fmt.Errorf("--server flag required on parent command")
	}

	// Load first-pass results
	suggestions, err := loadSuggestionsFromJSONL(pass2ResultsPath)
	if err != nil {
		return fmt.Errorf("load results: %w", err)
	}
	log.Printf("Loaded %d first-pass suggestions", len(suggestions))

	// Filter uncertain
	thresholds := map[string]bool{"low": true}
	if pass2Threshold == "medium" {
		thresholds["medium"] = true
	}
	var uncertain []map[string]interface{}
	for _, s := range suggestions {
		conf, _ := s["confidence"].(string)
		if thresholds[conf] {
			uncertain = append(uncertain, s)
		}
	}
	log.Printf("Found %d uncertain suggestions (%s and below)", len(uncertain), pass2Threshold)

	if len(uncertain) == 0 {
		log.Println("No uncertain suggestions to enrich.")
		return nil
	}

	// Load groups
	groupsData, err := os.ReadFile(pass2GroupsPath)
	if err != nil {
		return fmt.Errorf("read groups: %w", err)
	}
	var groups []json.RawMessage
	if err := json.Unmarshal(groupsData, &groups); err != nil {
		return fmt.Errorf("parse groups: %w", err)
	}

	// Collect author IDs for book enrichment
	authorIDs := collectAuthorIDsFromGroups(uncertain, groups)

	// Fetch book data
	log.Printf("Fetching books for %d authors from %s...", len(authorIDs), benchServerURL)
	booksByAuthor, err := fetchBooksForAuthorIDs(benchServerURL, authorIDs)
	if err != nil {
		return fmt.Errorf("fetch books: %w", err)
	}

	// Build enriched items
	var enriched []map[string]interface{}
	for _, s := range uncertain {
		gi, _ := s["group_index"].(float64)
		idx := int(gi)
		if idx < 0 || idx >= len(groups) {
			continue
		}

		var group map[string]interface{}
		json.Unmarshal(groups[idx], &group)

		evidence := buildBookEvidence(group, booksByAuthor)

		enriched = append(enriched, map[string]interface{}{
			"group_index":         idx,
			"original_group":      group,
			"previous_suggestion": s,
			"book_evidence":       evidence,
		})
	}

	// Create run directory
	ts := time.Now().Format("2006-01-02T15-04-05")
	runDir := filepath.Join(benchOutputDir, ts+"-pass2")
	if err := os.MkdirAll(runDir, 0775); err != nil {
		return err
	}

	userContent, _ := json.Marshal(enriched)
	_ = writeJSON(filepath.Join(runDir, "enriched_input.json"), enriched)

	// Submit batch
	return submitSingleBatch(cmd.Context(), apiKey, pass2Model, pass2SystemPrompt,
		"Review these uncertain suggestions with the new book title evidence:\n\n"+string(userContent),
		fmt.Sprintf("pass2_%s_%s", pass2Model, ts), runDir)
}

// loadSuggestionsFromJSONL loads suggestions from a batch_output.jsonl file.
func loadSuggestionsFromJSONL(path string) ([]map[string]interface{}, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var suggestions []map[string]interface{}
	for _, line := range splitJSONLines(data) {
		var resp struct {
			Response struct {
				StatusCode int             `json:"status_code"`
				Body       json.RawMessage `json:"body"`
			} `json:"response"`
		}
		if json.Unmarshal(line, &resp) != nil || resp.Response.StatusCode != 200 {
			continue
		}

		var completion struct {
			Choices []struct {
				Message struct {
					Content string `json:"content"`
				} `json:"message"`
			} `json:"choices"`
		}
		if json.Unmarshal(resp.Response.Body, &completion) != nil || len(completion.Choices) == 0 {
			continue
		}

		content := extractJSONFromContent(completion.Choices[0].Message.Content)
		var parsed struct {
			Suggestions []map[string]interface{} `json:"suggestions"`
		}
		if json.Unmarshal([]byte(content), &parsed) == nil {
			suggestions = append(suggestions, parsed.Suggestions...)
		}
	}
	return suggestions, nil
}

func collectAuthorIDsFromGroups(uncertain []map[string]interface{}, groups []json.RawMessage) map[int]bool {
	ids := map[int]bool{}
	for _, s := range uncertain {
		gi, _ := s["group_index"].(float64)
		idx := int(gi)
		if idx < 0 || idx >= len(groups) {
			continue
		}
		var group struct {
			Canonical struct {
				ID int `json:"id"`
			} `json:"canonical"`
			Variants []struct {
				ID int `json:"id"`
			} `json:"variants"`
		}
		if json.Unmarshal(groups[idx], &group) == nil {
			if group.Canonical.ID > 0 {
				ids[group.Canonical.ID] = true
			}
			for _, v := range group.Variants {
				if v.ID > 0 {
					ids[v.ID] = true
				}
			}
		}
	}
	return ids
}

func buildBookEvidence(group map[string]interface{}, booksByAuthor map[int][]BookInfo) map[string][]string {
	evidence := map[string][]string{}
	if canonical, ok := group["canonical"].(map[string]interface{}); ok {
		if id, ok := canonical["id"].(float64); ok {
			name, _ := canonical["name"].(string)
			if books, ok := booksByAuthor[int(id)]; ok {
				var titles []string
				for _, b := range books {
					t := b.Title
					if b.Series != "" {
						t += " (" + b.Series + ")"
					}
					titles = append(titles, t)
				}
				evidence[name] = titles
			}
		}
	}
	if variants, ok := group["variants"].([]interface{}); ok {
		for _, vi := range variants {
			v, ok := vi.(map[string]interface{})
			if !ok {
				continue
			}
			if id, ok := v["id"].(float64); ok {
				name, _ := v["name"].(string)
				if books, ok := booksByAuthor[int(id)]; ok {
					var titles []string
					for _, b := range books {
						t := b.Title
						if b.Series != "" {
							t += " (" + b.Series + ")"
						}
						titles = append(titles, t)
					}
					evidence[name] = titles
				}
			}
		}
	}
	return evidence
}

// BookInfo holds a book title and series for enrichment.
type BookInfo struct {
	Title  string
	Series string
}

func fetchBooksForAuthorIDs(serverURL string, authorIDs map[int]bool) (map[int][]BookInfo, error) {
	httpClient := newInsecureClient()

	books := map[int][]BookInfo{}
	offset := 0
	pageSize := 1000
	for {
		url := fmt.Sprintf("%s/api/v1/audiobooks?limit=%d&offset=%d", serverURL, pageSize, offset)
		resp, err := httpClient.Get(url)
		if err != nil {
			return nil, err
		}
		var page struct {
			Items []json.RawMessage `json:"items"`
		}
		json.NewDecoder(resp.Body).Decode(&page)
		resp.Body.Close()

		for _, raw := range page.Items {
			var book struct {
				Title    string `json:"title"`
				Series   string `json:"series"`
				AuthorID int    `json:"author_id"`
			}
			if json.Unmarshal(raw, &book) != nil {
				continue
			}
			if authorIDs[book.AuthorID] && len(books[book.AuthorID]) < 10 {
				books[book.AuthorID] = append(books[book.AuthorID], BookInfo{Title: book.Title, Series: book.Series})
			}
		}

		if len(page.Items) < pageSize {
			break
		}
		offset += pageSize
	}
	return books, nil
}

// newInsecureClient creates an HTTP client that skips TLS verification.
func newInsecureClient() *http.Client {
	return &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
		Timeout: 30 * time.Second,
	}
}

// submitSingleBatch submits a single request as a batch job.
func submitSingleBatch(ctx context.Context, apiKey, model, systemPrompt, userPrompt, customID, runDir string) error {
	clientOpts := []option.RequestOption{option.WithAPIKey(apiKey)}
	if baseURL := os.Getenv("OPENAI_BASE_URL"); baseURL != "" {
		clientOpts = append(clientOpts, option.WithBaseURL(baseURL))
	}
	client := openai.NewClient(clientOpts...)

	isReasoning := strings.HasPrefix(model, "o3") || strings.HasPrefix(model, "o4") || strings.HasPrefix(model, "o1") || strings.HasPrefix(model, "gpt-5")

	maxTokens := int64(16000)
	body := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": userPrompt},
		},
		"max_completion_tokens": maxTokens,
	}
	if !isReasoning {
		body["temperature"] = 0.0
		body["top_p"] = 1.0
		body["response_format"] = map[string]string{"type": "json_object"}
	}

	req := ai.BatchRequest{
		CustomID: customID,
		Method:   "POST",
		URL:      "/v1/chat/completions",
		Body:     body,
	}

	var buf bytes.Buffer
	line, _ := json.Marshal(req)
	buf.Write(line)
	buf.WriteByte('\n')

	_ = os.WriteFile(filepath.Join(runDir, "batch_input.jsonl"), buf.Bytes(), 0664)

	file, err := client.Files.New(ctx, openai.FileNewParams{
		File:    bytes.NewReader(buf.Bytes()),
		Purpose: openai.FilePurposeBatch,
	})
	if err != nil {
		return fmt.Errorf("upload file: %w", err)
	}

	batch, err := client.Batches.New(ctx, openai.BatchNewParams{
		InputFileID:      file.ID,
		Endpoint:         openai.BatchNewParamsEndpointV1ChatCompletions,
		CompletionWindow: openai.BatchNewParamsCompletionWindow24h,
	})
	if err != nil {
		return fmt.Errorf("create batch: %w", err)
	}

	jobInfo := []BatchJobInfo{{
		BatchID:     batch.ID,
		Config:      TestConfig{Model: model, PromptVariant: customID},
		Mode:        "pass2",
		NumChunks:   1,
		NumRequests: 1,
		InputFileID: file.ID,
		RunDir:      runDir,
	}}
	_ = writeJSON(filepath.Join(runDir, "batch_jobs.json"), jobInfo)

	log.Printf("Submitted batch %s", batch.ID)
	log.Printf("Check: ./audiobook-organizer dedup-bench check %s", runDir)
	return nil
}

File: scripts/dedup_bench_pass2.py
#!/usr/bin/env python3
"""Second-pass enrichment for uncertain dedup suggestions.

Takes medium/low confidence results from a first-pass groups run,
enriches them with book title data from the server, and resubmits
for a more informed decision.

Usage:
    python3 scripts/dedup_bench_pass2.py \
        --server https://172.16.2.30:8484 \
        --results testdata/dedup-bench/2026-03-07T20-11-15/runs/gpt-5.1_baseline_t0.0_groups/batch_output.jsonl \
        --groups testdata/dedup-bench/2026-03-07T20-11-15/groups.json \
        --model gpt-5.1

Loads API keys from scripts/.env automatically.
"""

import argparse
import json
import os
import ssl
import sys
import time
import urllib.request
from datetime import datetime
from pathlib import Path

# Load .env from scripts/ directory
env_path = Path(__file__).parent / ".env"
if env_path.exists():
    for line in env_path.read_text().splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            key, _, value = line.partition("=")
            os.environ.setdefault(key.strip(), value.strip())

from openai import OpenAI

SYSTEM_PROMPT = """You are an expert audiobook metadata reviewer performing a SECOND-PASS verification.

You previously reviewed author groups and flagged some as uncertain. Now you are given:
1. The original group of author name variants
2. Your previous suggestion and reasoning
3. NEW DATA: Book titles for each author variant, so you can determine if they are the same person

Use the book titles to make a more informed decision:
- If both variants have books in the same series or genre, they are likely the same person → merge
- If they have completely unrelated books, they are likely different people → split
- Narrators tend to have books across many unrelated genres/authors
- Publishers have many unrelated books

INITIALS FORMATTING: Always use spaces after periods in initials: "C. B. Lee" not "C.B. Lee".

Return ONLY valid JSON: {"suggestions": [{"group_index": N, "action": "merge|split|rename|skip|alias|reclassify", "canonical_name": "Correct Name", "reason": "brief explanation using book title evidence", "confidence": "high|medium|low", "roles": {"author": {"name": "Name", "variants": ["V1"], "reason": "why"}, "narrator": {"name": "Name", "ids": [], "reason": "why"}, "publisher": {"name": "Name", "ids": [], "reason": "why"}}}]}"""


def fetch_books_for_authors(server_url: str, author_ids: list) -> dict:
    """Fetch book titles for specific author IDs."""
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE

    books_by_author = {}
    # Fetch all books (paginated) and filter
    offset = 0
    page_size = 1000
    while True:
        url = f"{server_url}/api/v1/audiobooks?limit={page_size}&offset={offset}"
        req = urllib.request.Request(url)
        with urllib.request.urlopen(req, context=ctx) as resp:
            page = json.loads(resp.read())

        items = page.get("items", [])
        for book in items:
            aid = book.get("author_id", 0)
            if aid in author_ids:
                if aid not in books_by_author:
                    books_by_author[aid] = []
                books_by_author[aid].append(
                    {
                        "title": book.get("title", ""),
                        "series": book.get("series", ""),
                    }
                )

        if len(items) < page_size:
            break
        offset += page_size

    return books_by_author


def main():
    parser = argparse.ArgumentParser(
        description="Second-pass enrichment for uncertain suggestions"
    )
    parser.add_argument("--server", required=True, help="Server URL")
    parser.add_argument(
        "--results", required=True, help="Path to first-pass batch_output.jsonl"
    )
    parser.add_argument(
        "--groups", required=True, help="Path to groups.json from first pass"
    )
    parser.add_argument(
        "--model", default="gpt-5.1", help="Model to use for second pass"
    )
    parser.add_argument(
        "--threshold",
        default="medium",
        choices=["medium", "low"],
        help="Include suggestions at this confidence and below (default: medium = medium+low)",
    )
    parser.add_argument(
        "--output", default="testdata/dedup-bench", help="Output base directory"
    )
    parser.add_argument(
        "--dry-run", action="store_true", help="Build prompts but don't submit"
    )
    args = parser.parse_args()

    api_key = os.environ.get("OPENAI_API_KEY", "")
    if not api_key and not args.dry_run:
        print("ERROR: OPENAI_API_KEY not set (check scripts/.env)")
        sys.exit(1)

    # Load first-pass results
    results_path = Path(args.results)
    suggestions = []
    for line in results_path.read_text().strip().split("\n"):
        resp = json.loads(line)
        if resp.get("response", {}).get("status_code") != 200:
            continue
        content = resp["response"]["body"]["choices"][0]["message"]["content"]
        parsed = json.loads(content)
        suggestions.extend(parsed.get("suggestions", []))

    # Filter uncertain ones
    thresholds = ["low"] if args.threshold == "low" else ["medium", "low"]
    uncertain = [s for s in suggestions if s.get("confidence") in thresholds]
    print(
        f"First pass: {len(suggestions)} total, {len(uncertain)} uncertain ({', '.join(thresholds)})"
    )

    if not uncertain:
        print("No uncertain suggestions to enrich. Done.")
        return

    # Load groups
    groups = json.loads(Path(args.groups).read_text())

    # Collect all author IDs we need books for
    all_author_ids = set()
    for s in uncertain:
        gi = s.get("group_index", -1)
        if 0 <= gi < len(groups):
            group = groups[gi]
            # Extract author IDs from the group
            canonical = group.get("canonical", {})
            if canonical.get("id"):
                all_author_ids.add(canonical["id"])
            for v in group.get("variants", []):
                if v.get("id"):
                    all_author_ids.add(v["id"])

    print(f"Fetching books for {len(all_author_ids)} authors...")
    books_by_author = fetch_books_for_authors(args.server, all_author_ids)
    total_books = sum(len(b) for b in books_by_author.values())
    print(f"  Found {total_books} books across {len(books_by_author)} authors")

    # Build enriched prompts — one request per uncertain suggestion
    enriched_items = []
    for s in uncertain:
        gi = s.get("group_index", -1)
        if gi < 0 or gi >= len(groups):
            continue

        group = groups[gi]
        canonical = group.get("canonical", {})
        variants = group.get("variants", [])

        # Build book evidence
        evidence = {}
        if canonical.get("id"):
            aid = canonical["id"]
            titles = books_by_author.get(aid, [])
            evidence[canonical.get("name", f"ID {aid}")] = [
                t["title"] + (f" ({t['series']})" if t.get("series") else "")
                for t in titles[:10]
            ]
        for v in variants:
            if v.get("id"):
                aid = v["id"]
                titles = books_by_author.get(aid, [])
                evidence[v.get("name", f"ID {aid}")] = [
                    t["title"] + (f" ({t['series']})" if t.get("series") else "")
                    for t in titles[:10]
                ]

        enriched_items.append(
            {
                "group_index": gi,
                "original_group": {
                    "canonical": {
                        "name": canonical.get("name", ""),
                        "id": canonical.get("id"),
                    },
                    "variants": [
                        {"name": v.get("name", ""), "id": v.get("id")} for v in variants
                    ],
                },
                "previous_suggestion": {
                    "action": s.get("action"),
                    "canonical_name": s.get("canonical_name"),
                    "confidence": s.get("confidence"),
                    "reason": s.get("reason"),
                },
                "book_evidence": evidence,
            }
        )

    # Build the batch request
    user_content = json.dumps(enriched_items, indent=2)
    print(
        f"\nEnriched payload: {len(enriched_items)} items, ~{len(user_content) // 4} tokens"
    )

    # Create run directory
    ts = datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
    run_dir = Path(args.output) / f"{ts}-pass2"
    run_dir.mkdir(parents=True, exist_ok=True)

    # Save enriched data
    (run_dir / "enriched_input.json").write_text(json.dumps(enriched_items, indent=2))
    (run_dir / "config.json").write_text(
        json.dumps(
            {
                "model": args.model,
                "mode": "pass2-enrichment",
                "first_pass_results": str(args.results),
                "threshold": args.threshold,
                "uncertain_count": len(enriched_items),
            },
            indent=2,
        )
    )

    if args.dry_run:
        print(f"Dry run — enriched data saved to {run_dir}")
        print(f"\nSample enriched item:")
        print(json.dumps(enriched_items[0], indent=2))
        return

    # Submit as batch
    is_reasoning = args.model.startswith("o") or args.model.startswith("gpt-5")

    body = {
        "model": args.model,
        "messages": [
            {"role": "system", "content": SYSTEM_PROMPT},
            {
                "role": "user",
                "content": f"Review these uncertain suggestions with the new book title evidence:\n\n{user_content}",
            },
        ],
        "max_completion_tokens": 16000,
    }

    if not is_reasoning:
        body["temperature"] = 0.0
        body["top_p"] = 1.0
        body["response_format"] = {"type": "json_object"}

    req_line = (
        json.dumps(
            {
                "custom_id": f"pass2_{args.model}_{ts}",
                "method": "POST",
                "url": "/v1/chat/completions",
                "body": body,
            }
        )
        + "\n"
    )

    # Save JSONL
    (run_dir / "batch_input.jsonl").write_text(req_line)

    client = OpenAI(api_key=api_key)

    print(f"Uploading and submitting batch job...")
    file = client.files.create(file=req_line.encode(), purpose="batch")
    batch = client.batches.create(
        input_file_id=file.id,
        endpoint="/v1/chat/completions",
        completion_window="24h",
    )

    job_info = {
        "batch_id": batch.id,
        "input_file_id": file.id,
        "model": args.model,
        "mode": "pass2-enrichment",
        "uncertain_count": len(enriched_items),
    }
    (run_dir / "batch_info.json").write_text(json.dumps(job_info, indent=2))

    # Also save as batch_jobs.json for compatibility with check script
    (run_dir / "batch_jobs.json").write_text(
        json.dumps(
            [
                {
                    "batch_id": batch.id,
                    "config": {
                        "model": args.model,
                        "prompt_variant": "pass2-enrichment",
                    },
                    "mode": "pass2",
                    "run_dir": str(run_dir),
                }
            ],
            indent=2,
        )
    )

    print(f"\nSubmitted batch job: {batch.id}")
    print(f"Output: {run_dir}")
    print(f"Check: python3 scripts/dedup_bench_check.py {run_dir}")


if __name__ == "__main__":
    main()
